### PR TITLE
Alluxio ui

### PR DESCRIPTION
**Pull request by WilliamZapata**
Replace the current Web UI.

### CODE REVIEW

File: alluxio-ui/common/src/components/FileView/FileView.test.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        'debug': true,
+        'fatalError': '',
+        'fileData': '',
+        'fileInfos': [],
+        'invalidPathError': '',
+        'ntotalFile': 0,
+        'viewingOffset': 0
+      }
+    };
+  });
```
Comment (calvinjia) [inline alluxio-ui/common/src/components/FileView/FileView.test.tsx]: Does it make sense to test with non empty data as well?

File: alluxio-ui/common/src/utilities/handlers/getDropdownItemHandler.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+import React from 'react';
+
+import {FieldValidatorFunctionsType, getDropdownItemValidator} from './getDropdownItemValidator';
+
+type DropdownValidatorFunctionType = (this: any, event: React.FormEvent<HTMLInputElement>, isFormSubmission?: boolean) => void;
+
+export const getDropdownItemHandler = (elementId: string, elementName: string, fieldValidatorFunctions: FieldValidatorFunctionsType) => {
```
Comment (calvinjia) [inline alluxio-ui/common/src/utilities/handlers/getDropdownItemHandler.tsx]: Where is this used?

File: alluxio-ui/common/src/components/FileView/FileView.test.tsx (hunk 1 above)
Comment (WilliamZapata) [reply alluxio-ui/common/src/components/FileView/FileView.test.tsx]: Yes, absolutely, this is in the to-do pile

File: alluxio-ui/common/src/utilities/misc/bytesToString.tsx [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+export const bytesToString = (bytes: number) => {
+  if (bytes === 0) {
+    return '0Byte';
+  }
+  const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
+  const i = parseInt('' + Math.floor(Math.log(bytes) / Math.log(1024)), 10);
```
Comment (calvinjia) [inline alluxio-ui/common/src/utilities/misc/bytesToString.tsx]: Why do we convert from float(?) to string and then back to int? Is there a way to just do the math and convert to int?

File: alluxio-ui/common/src/components/Paginator/Paginator.tsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                              disabled={numLimit === 500}>View 500 rows</DropdownItem>
+              </DropdownMenu>
+            </UncontrolledButtonDropdown>
+          </PaginationItem>
+        </Pagination>
+      </div>
+    );
+  }
+
+  private getNumber(numString?: string) {
```
[the remark this answers is not in the document]
Comment (WilliamZapata) [reply alluxio-ui/common/src/components/Paginator/Paginator.tsx]: I think you are correct, I should probably remove the `if (!numString) {...}` part. Good catch.

[the remark this answers is not in the document]
Comment (WilliamZapata) [reply alluxio-ui/common/src/components/Paginator/Paginator.tsx]: I will check for numString being undefined and return null in the beginning.

File: alluxio-ui/common/src/utilities/misc/isExternalLink.tsx [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+export const isExternalLink = (url: string | undefined) => !!url && !/^\/[^/].*/.test(url);
```
Comment (calvinjia) [inline alluxio-ui/common/src/utilities/misc/isExternalLink.tsx]: How does this determine if it is an external link?

File: core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java [hunk 6]
```diff
@@ -126,23 +201,783 @@ public Response getInfo(@QueryParam(QUERY_RAW_CONFIGURATION) final Boolean rawCo
...
+            throw new IOException(e);
+          }
+          response.setViewingOffset(offset);
+          return response;
+        }
+
+        if (currentPath.isRoot()) {
+          response.setPathInfos(new UIFileInfo[0]);
+        } else {
+          String[] splitPath = PathUtils.getPathComponents(currentPath.toString());
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java]: There is a bug in this section of the code. We need to create a new variable, ie. `pathComponent = new AlluxioURI(AlluxioURI.SEPARATOR)` to build the path components. If we tack on to the current path, we will get incorrect paths (for example, for `/a/b/c`, we want `/a`, `/a/b`, the current implementation will return `/a/b/c/a` and `/a/b/c/a/b`)

Comment (WilliamZapata) [reply core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java]: oh ok, will fix

**Comment by WilliamZapata** [comment]
I agree, there is also the download servlet that should be ported over to a restful api.

File: core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java [hunk 7]
```diff
@@ -456,8 +458,9 @@ public Response getWebUIBrowse(@DefaultValue("/") @QueryParam("path") String req
           UIFileInfo[] pathInfos = new UIFileInfo[splitPath.length - 1];
           fileId = mFileSystemMaster.getFileId(currentPath);
           pathInfos[0] = new UIFileInfo(mFileSystemMaster.getFileInfo(fileId));
-          for (int i = 1; i < splitPath.length - 1; i++) {
-            currentPath = currentPath.join(splitPath[i]);
+          for (int i = splitPath.length - 1; i == 0; i++) {
+            String[] pathArray = Arrays.copyOfRange(splitPath, 0, splitPath.length - i);
+            currentPath = new AlluxioURI(StringUtils.join(pathArray));
```
Comment (calvinjia) [inline core/server/master/src/main/java/alluxio/master/meta/AlluxioMasterRestServiceHandler.java]: I wouldn't mutate `currentPath` here, could we instead do something like

```
AlluxioURI breadcrumb = new AlluxioURI(AlluxioURI.SEPARATOR);
for (...) {
  breadcrumb = breadcrumb.join(splitPath[i]);
  fileId = mFileSystemMaster.getFileId(breadcrumb);
  ...
}  
```

**Pull request merged**